### PR TITLE
Implement virtual functions, allow including interfaces, make GObject subclasses GC-safe

### LINKED ISSUE #10: Add ability to implement interfaces (and their methods)
It would be useful to be able to implement interfaces (and their methods)
My use-case is to implement `Gdk::Paintable` and its `snapshot` method.

### PR DESCRIPTION
**Pull request by BlobCodes**
Closes #10 

This PR allows the user to override virtual functions and allows including interfaces (and overriding their virtual functions too).

This is done using a macro, which automagically adds boilerplate to the `_class_init` method when a newly added method starts with "do_".

Here's an example of how this looks:
```crystal
require "gtk4"

class TestWidget < Gtk::Widget
  @test = "hi"

  def do_size_allocate(width, height, baseline)
    puts "ALLOC: #{width}x#{height}, baseline: #{baseline}, test: #{@test}, self: #{self}"
  end
end

class TestPaintable < GObject::Object
  include Gdk::Paintable

  @test = "hi"

  def do_snapshot(snapshot, width, height)
    puts "Snapshot: #{width}x#{height}, snapshot: #{snapshot}, test: #{@test}, self: #{self}"
  end
end

class TestPaintableChild < TestPaintable
  @test2 = "hi2"

  def do_snapshot(snapshot, width, height)
    puts "Snapshot: #{width}x#{height}, snapshot: #{snapshot}, test: #{@test}, test2: #{@test2}"
  end
end

app = Gtk::Application.new("test.test", Gio::ApplicationFlags::None)
app.activate_signal.connect do
  window = Gtk::ApplicationWindow.new(app)
  box = Gtk::Box.new
  box.append TestWidget.new
  box.append Gtk::Picture.new_for_paintable(TestPaintable.new)
  box.append Gtk::Picture.new_for_paintable(TestPaintableChild.new)
  box.append Gtk::Picture.new_for_paintable(TestPaintable.new)
  window.child = box
  window.present
end
exit(app.run)
```

This code works and shows no error messages from gtk:
Bildschirmfoto vom 2022-06-08 22-22-16

Besides this test application, there are no tests yet. I have no clue how one would add tests for this, but I'll try to add some if it's required.

`make test` runs without problems.

EDIT: replaced `vfunc_` with `do_`

### CODE REVIEW

**Comment by BlobCodes** [comment]
The interfaces are just added in crystal's main. I originally wanted to do that in `_class_init` using the way you showed in [this comment](https://github.com/hugopl/gi-crystal/issues/10#issuecomment-1119212219), but then I got error messages like these:
```
(crystal-run-test2.tmp:64685): GLib-GObject-WARNING **: 21:32:07.846: attempting to add an interface (GdkPaintable) to class (TestPaintable) after class_init
```

**Comment by BlobCodes** [comment]
One thing that probably should be changed is that in `v_func.ecr` I directly used `g_object_get_qdata` and not `<%= object.qdata_get_func %>` or `self.new`.

`self.new(pointer, transfer)` does not work because it is defined by gi-crystal and not in any macros:
```
Error: method TestPaintable.new must return TestPaintable+ but it is returning GObject::Object+
```

and I have no access to `object.qdata_get_func` because when the virtual function is defined on an interface, I have no access to it.

**Comment by BlobCodes** [comment]
I should also note that you have to re-include a interface if you want to implement methods from an interface but the parent class does not implement it.

Example:
```
class TestPaintableChild < TestPaintable
  include Gdk::Paintable
  @test2 = "hi2"

  def vfunc_get_intrinsic_aspect_ratio
    1.0
  end
end
```

I have to find a way to show an error or automatically solve this because right now this compiles but silently does not work.

**Comment by hugopl** [comment]
Awesome! I still need to do a more detailed review, but move the `GObject::Object` code away from `object.ecr` was really a good change. Would be awesome to land this change separately to easy the review, it's boring work, I know, so if you don't mind I could do this on your behalf.

I think we could follow PyGObject, and let the vfuncs be named `do_#{vfunc.name}` instead of `vfunc_#{vfunc.name}`.

Have tests for this is very important, currently the project has a [small dummy C library](https://github.com/hugopl/gi-crystal/tree/master/spec/libtest) where we create a binding for it and the tests are written against this binding. So to create a test we need first to add a virtual function on some object of libtest. I have some draft for this that I made weeks ago, I just [squashed and pushed the commit](https://github.com/hugopl/gi-crystal/commit/d0d54e3583d1c40431f78dfc79c3dc251a23d75c#diff-45a5ddb882dc587ac7a1abdbfae7c6caaa8a3219ff2fa23d733598dae5937414) to a branch now, so maybe it can be useful 

I can also help with these tests.

I'll be travelling next days, so I'll only be able to check this out on Monday :-/.

**Comment by hugopl** [comment]
> One thing that probably should be changed is that in v_func.ecr I directly used g_object_get_qdata and not <%= object.qdata_get_func %> or self.new.

This is not a problem, `<%= object.qdata_get_func %>` is used because of `GObject::ParamSpec` object, that is a object registered in GObject type system but isn't in the `GObject::Object` tree, i.e. it's a fundamental object as GLib calls it.

As only `GObject::Object` can have interfaces, this is not a problem... and keep things simpler.

**Comment by hugopl** [comment]
> I should also note that you have to re-include a interface if you want to implement methods from an interface but the parent class does not implement it.

Maybe using the `finished` macro can help, but I need to look in more detail the generated code.

**Comment by hugopl** [comment]
> The interfaces are just added in crystal's main. I originally wanted to do that in `_class_init` using the way you showed in [this comment](https://github.com/hugopl/gi-crystal/issues/10#issuecomment-1119212219), but then I got error messages like these:
> 
> ```
> (crystal-run-test2.tmp:64685): GLib-GObject-WARNING **: 21:32:07.846: attempting to add an interface (GdkPaintable) to class (TestPaintable) after class_init
> ```

This is because `self._class_init` calls `previous_def` at end, so everything goes after it initialize the class in gobject type system... we could add a `self._class_pre_init`, called at the start of `self._class_init` with a empty default implementation... that the interfaces could re-implement calling `previous_def`, creating a the _previous_def chain_ like I tried in that comment on the issue.

File: ecr/v_func.ecr [hunk 1]
```diff
@@ -0,0 +1,20 @@
+    private macro _register_vfunc_<%= vfunc.name %>
+      private def self._vfunc_<%= vfunc.name %>(_this : Pointer(Void), <% generate_lib_args(io, vfunc) %>) : <%= return_type %>
+        <%- write_implementations(io) -%>
+        <%-= '\n' unless vfunc.args.empty? -%>
```
Comment (hugopl) [inline ecr/v_func.ecr:4]: Better to just add an extra line break and let `crystal tool format` do the job, so there's less noise in the .ecr file.

File: ecr/v_func.ecr [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        LibGObject
+          .g_object_get_qdata(_this.unsafe_as(Pointer(Void)), GICrystal::INSTANCE_QDATA_KEY)
```
Comment (hugopl) [inline ecr/v_func.ecr:6]: `_this` is already a `Pointer(Void)`
```suggestion
          .g_object_get_qdata(_this, GICrystal::INSTANCE_QDATA_KEY)
```

**Review by hugopl — CHANGES_REQUESTED** [review]

File: ecr/v_func.ecr (hunk 1 above)
Comment (BlobCodes) [reply ecr/v_func.ecr:4]: This is inside a macro, `crystal tool format` will not format anything here.
Here's an example what will be generated:
```crystal
    private macro _register_vfunc_get_intrinsic_width
      private def self._vfunc_get_intrinsic_width(_this : Pointer(Void), ) : Int32
        LibGObject
          .g_object_get_qdata(_this.unsafe_as(Pointer(Void)), GICrystal::INSTANCE_QDATA_KEY)
          .as(self)
          .vfunc_get_intrinsic_width()
      end

      def self._install_iface_Gdk__Paintable(type_struct : Pointer(LibGObject::TypeInterface)) : Nil
        vfunc_ptr = (type_struct.as(Pointer(Void)) + 40).as(Pointer(Pointer(Void)))
        vfunc_ptr.value = (->_vfunc_get_intrinsic_width(Pointer(Void))).pointer
        previous_def
      end
    end

    private macro _register_vfunc_snapshot
      private def self._vfunc_snapshot(_this : Pointer(Void), lib_snapshot :  Pointer(Void), lib_width :  Float64, lib_height :  Float64, ) : Void
# Generator::BuiltInTypeArgPlan
snapshot=Gdk::Snapshot.new(lib_snapshot, :none)
width=lib_width
height=lib_height

        LibGObject
          .g_object_get_qdata(_this.unsafe_as(Pointer(Void)), GICrystal::INSTANCE_QDATA_KEY)
          .as(self)
          .vfunc_snapshot(snapshot, width, height)
      end

      def self._install_iface_Gdk__Paintable(type_struct : Pointer(LibGObject::TypeInterface)) : Nil
        vfunc_ptr = (type_struct.as(Pointer(Void)) + 16).as(Pointer(Pointer(Void)))
        vfunc_ptr.value = (->_vfunc_snapshot(Pointer(Void), Pointer(Void), Float64, Float64)).pointer
        previous_def
      end
    end
```

Notice the comma at the end of `_vfunc_get_intrinsic_width`'s arguments and the implementations having no identation (after `crystal tool format`)

**Comment by BlobCodes** [comment]
This commit will always include all interfaces in all inherited user types, meaning that you can implement methods from interfaces you have only included in the parent class:

```crystal
class TestPaintable < GObject::Object
  include Gdk::Paintable

  def do_snapshot(snapshot, width, height)
    puts "Snapshot: #{width}x#{height}, snapshot: #{snapshot}, self: #{self}"
  end
end

class TestPaintableChild < TestPaintable
  def do_get_intrinsic_aspect_ratio
    1.0
  end
end
```

Using this approach means that if you include an interface multiple times, you get a warning on the terminal:
```
(crystal-run-test2.tmp:37338): GLib-GObject-WARNING **: 16:24:06.843: cannot add interface type 'GdkPaintable' to type 'TestPaintable', since type 'TestPaintable' already conforms to interface
```

I could fix this using a class-variable, like this:
```crystal
@@_implemented_interface_{{namespace_name}}__{{type_name}} = false

def self._install_ifaces
  unless @@_implemented_interface_{{namespace_name}}__{{type_name}}
    # (do stuff)
    @@_implemented_interface_{{namespace_name}}__{{type_name}} = true
  end
  previous_def
end
```

..but I don't know if it's worth the overhead when the user can just remove the duplicated `include` statement.
Wdyt @hugopl ?

**Comment by BlobCodes** [comment]
All commits after 1d8b4b907c67a21b780b95c0ad208edf7d8ac30d are my progress on allowing to create gobjects which do not need to be referenced in crystal. My plan is to split a gobject subclass into two classes - an ordinary pointer (like all other gobjects) and a private data class. A reference to the private data class (which would be allocated using malloc_uncollectable and manually freed using the finalize vfunction) would be placed into the gobject struct.

However, I don't know if that would make this PR too big. Should I rather split it up into multiple PRs? (and is this a suitable way to address the issue?)

File: src/bindings/g_object/object.cr [hunk 3]
```diff
@@ -7,6 +7,14 @@ module GObject
   class Object
     macro inherited
       {% unless @type.annotation(GObject::GeneratedWrapper) %}
+        macro method_added(method)
+          {% verbatim do %}
+            {% if method.name.starts_with?("do_") %}
```
Comment (hugopl) [inline src/bindings/g_object/object.cr:12]: This is too generic and can cause problems for user code writing methods named `do_something` when there's no `something` function.

Would be nice to have a predefined list of virtual methods here, but to achieve this this macro need to go to `object.ecr`

another option is to have a `Virtual` annotation, but IMO the less annotations needed the better.

**Comment by hugopl** [comment]
> However, I don't know if that would make this PR too big.

Yes it is, we can split it to first do the basics, the optimize things.

What problems you had with `GC_COLLECTED_QDATA_KEY` to need to add `src/bindings/g_object/gc.cr`?

**Comment by BlobCodes** [comment]
One disadvantage of this design is that the instance variables of an object implementation are only availabe inside it (TestPaintableChild cannot access `@test`, you need to define a protected property).

However after ***lots*** of testing I think this is the best design that allows you to have references to the crystal world.

**Comment by hugopl** [comment]
I think is better to go simpler and wait until a better approach appear.

If I remember correctly the problem is (Please correct me if I'm wrong):

_The Crystal object was garbage collected, but the C object is alive, then someone call a virtual function and if the virtual function implemented in Crystal uses any object attribute it will crash._ 

If the problem is just that, I believe that raising `GICrystal::ObjectCollectedError` is ok for now.

So we can do things by steps, this MR could initially just do:

- Register interface in user defined types, a.k.a. let users use interfaces.
- Register virtual functions, a.k.a. let users implement virtual functions defined in interfaces.

If the Crystal object was collected by GC, an `GICrystal::ObjectCollectedError` is raised.

Ideally the virtual functions could be separated from the interface implementation, since they are two different things, but is ok to have both in a single MR.

**Comment by BlobCodes** [comment]
> I think is better to go simpler and wait until a better approach appear.
> 
> If I remember correctly the problem is (Please correct me if I'm wrong):
> 
> The Crystal object was garbage collected, but the C object is alive, then someone call a virtual function and if the virtual function implemented in Crystal uses any object attribute it will crash.
> 
> If the problem is just that, I believe that raising GICrystal::ObjectCollectedError is ok for now.

I don't agree, I think custom subclasses are not very useful if you need to reference them in crystal.
But I guess I'll first split this PR in two, so we can have separate discussions about the two.

As far as I can see, we only have three options:
- Raise if the object was collected by the GC (current, bad)
- Store all data in an object allocated using `GC_malloc_uncollectable` (my approach)
- Only allow GObject properties to be stored in a GObject subclass (I think GJS is doing this, performant but more complicated)

**Comment by hugopl** [comment]
> I think custom subclasses are not very useful if you need to reference them in crystal.

I don't think so, off course is better to not have this restriction. But I think we must solve one problem at a time, so will be easier to find better solutions.

- Raise an exception is the easier solution, not the better one.
- Be able to declare and use GObject properties is another one, also not really the better solution since it requires the vfunc to only use the object properties, not normal Crystal instance variables, and forbid users from declaring instance variables looks weird to me.

The ideal solution would be something that keep the programmer the feeling that he/she still programming in Crystal, as natural as possible, this is why I didn't like the approach of including all class declaration under a `implementation` macro. So I think is better to deliver the vfuncs raising the exceptions now, then think more on a better solution.

**Comment by BlobCodes** [comment]
> The ideal solution would be something that keep the programmer the feeling that he/she still programming in Crystal, as natural as possible, this is why I didn't like the approach of including all class declaration under a implementation macro.

This is the implementation macro:
```crystal
macro implementation(&block)
  @[GObject::PrivateDataClass]
  class PrivateData < self
    {{yield}}
  end
end
```

We could also just make the developer define a second class itself.
I don't think it's unnatural - the developer knows there's a pointer class and a data class. The data class inherits from the pointer class and has an annotation.

**Comment by BlobCodes** [comment]
I now split this PR in #26 and #28, so we should continue discussions there.

**Pull request closed without merging**